Update CreateDashboard Script

ardelato opened the pull request:
## Description

There have been quite a few changes to the underlying core logic. This has affected the metrics being sent to DataDog which in turn means the dashboards created through the `createDashboard` script are now outdated and ineffective.

The `createDashboard` script has been revised in this update to enhance the efficiency and functionality of the dashboards it generates.

| Before | After |
|--------|--------|
| ![image](https://github.com/iFixit/vigilo/assets/22064420/fc5db3b0-d56e-41d0-9191-7bae62217011) | ![image](https://github.com/iFixit/vigilo/assets/22064420/cb4dc22f-351e-41b9-a023-7feeef9baccb) | 

### CR

This mainly adds a new `Query Value` widget to the dashboard building logic. The new `Query Value` widget provides an average for all URLs of each page type, facilitating quick issue identification and correlation with data points. In contrast, the `Timeseries` widget is more useful at tracking changes over time for each URL. Additionally, the 
`Timeseries` widget has been enhanced to easily incorporate `customLinks`, allowing for direct access to relevant pages associated with the data points.

### QA

- [ ] Confirm running `pnpm create-dashboard` now creates a dashboard that shows the data properly
- [ ] Confirm for the `Timeseries` widget the datapoints have a `Visit Webpage` link that will take you to the actual webpage that was audited.

Closes: https://github.com/iFixit/ifixit/issues/51901

ardelato committed:
fix(createDashboard): use env var for host

We have added the feature of being able to run `vigilo` in a Kubernetes
cluster which means the host tag sent to DataDog will be 'docker-container'.
Therefore, let's use an environment variable to configure the host query
parameter.

ardelato committed:
feat(createDashboard): add alerts for query value widget

We will be adding a Query Value widget to the dashboard so we will need to
add alerts to said widget. Let's differentiate between the Timeseries
and Query Value alerts and add the new Query Value alerts

ardelato committed:
feat(createDashboard): create request for the query value widget

Similar to the timeseries request function, we will create a new function to
handle the creation of requests for the query value widget. One thing to note,
unlike the timeseries widget, the query value widget implements the alerts
through the requests definition.

ardelato committed:
fix(createDashboard): fix the queries for the widgets

We've updated the lighthouse metrics to send both a 'score' and 'value' datapoint
for each audit. As such we need to update the queries to specify the new
subpath. For most of the audits, the 'value' datapoint is more useful than the
'score' datapoint. The only audits that we would use the 'score' datapoint are
those with null 'value' datapoints like the 'is_crawlable' audit.

ardelato committed:
feat(createDashboard): add query value widget function

Similar to the Timeseries widget function, this adds a new function to create
a Query Value widget.

ardelato committed:
feat(createDashboard): add layout parameter for widgets

We will need to pair up the query value widget and timeseries widget together
in the same row. On top of that, we don't want the widgets to expand the full
width. Therefore, we will pass a parameter to modify the size and position
of the widgets.

ardelato committed:
feat(createDashboard): add query value widgets to iteration

This modifies the createWidgetForAllPageTypes function to create a query value
widget at the same time as the timeseries widget. On top of that, we will
keep track of the position of the widgets to layout them out properly per
group.

ardelato committed:
feat(createDashboard): add custom link for timeseries

This adds a custom link to the timeseries widgets such that we can link the
actual webpage for the datapoint being displayed. This will make it convenient
to debug and investigate the data without having to manually type out the URL.

deltuh-vee commented:
QA 🎬 
- [x] Running pnpm create-dashboard now creates a dashboard that shows the data properly. 
- [x] For the Timeseries widget, the datapoints have a Visit Webpage link that will take you to the actual webpage that was audited.

ardelato committed:
fix(createDashboard): fix comparator value units

For some of the audits, we weren't scaling the comparator value to the correct
unit types of the metrics being sent. Most of the metrics are in milliseconds,
so we need to scale the comparator value to milliseconds as well as these values
assumed the metric was in seconds.

ardelato committed:
fix(createDashboard): fix alert display type naming convention

The alert display type is CamelCase, not snake_case. This caused the warning
marker to not have the correct display type.

kthaler commented:
QA 💎 - Running `pnpm create-dashboard` creates a dashboard with the same data tables and format as the one in the PR description.

deploy_block 🦖 - I am not seeing the "Visit Webpage" link on the datapoints in a Timeseries widget:

<img width="497" alt="Screenshot 2024-02-14 at 5 36 10 PM" src="https://github.com/iFixit/vigilo/assets/145375039/33a5b132-ec79-45e6-b3b1-edd215296a1a">

ardelato commented:
un_deploy_block ⚡ 
@kthaler was that graph manually added? If not, what does your `lh-config.js` and `urls.json` files look like?

The other graphs within the groups do have the `Visit Webpage` links. 

![image](https://github.com/iFixit/vigilo/assets/22064420/02e8ce09-0ee3-4fc0-9ec8-faa002329f72)

kthaler commented:
un_deploy_block 🦕 
QA 💎 - The data points on the Timeseries widget have a link to the webpage that was audited.

mlahargou commented on src/scripts/createDashboard.ts:
```diff
@@ -42,22 +42,113 @@ const ALERT_MARKERS = {
 }
 
+function fetchQueryValueAlertMarkersForAudit(auditName: string) {
+    return ALERT_MARKERS_QUERY_VALUE.hasOwnProperty(auditName) ? ALERT_MARKERS_QUERY_VALUE[auditName] : []
```
I see us using `hasOwnProperty` in a few places. Can't we simplify this?

```
return ALERT_MARKERS_QUERY_VALUE[auditName] ?? []

ardelato replied:
TBH, this whole script could be improved and simplified. I thought about refactoring it by using a Factory design pattern but that is quite the over haul for the scope of the issue. 

https://github.com/iFixit/vigilo/compare/main...testing-factory-pattern-for-create-dashboards-script

mlahargou reviewed (approved):
I don't pretend to understand all of the queries. I think QA is more important there. 

The code seems sane. CR 📱
